File: 90. Subsets II.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        self.res = [[]]
        nums = sorted(nums)
        for k in xrange(1, len(nums)+1):
            self.combination(nums, [], 0, k)
        return self.res
    
    def combination(self, nums, path, start, k):
        if len(path) == k:
            self.res += [path[:]]
            return
        i = start
        while i<len(nums):
            self.combination(nums, path + [nums[i]], i+1, k)
            i += 1
            while i<len(nums) and nums[i] == nums[i-1]:
                i += 1

File: 90. Subsets II.cpp
class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        res.push_back({});
        sort(nums.begin(), nums.end());
        for (int k = 1; k <= nums.size(); k++)
            combination(nums, {}, 0, k);
        return res;
    }
    
    void combination(vector<int>& nums, vector<int> path, int start, int k)
    {
        if (path.size() == k)
            res.push_back(path);
        int i = start;
        while (i < nums.size())
        {
            path.push_back(nums[i]);
            combination(nums, path, i + 1, k);
            path.pop_back();
            i++;
            while (i < nums.size() && nums[i] == nums[i-1])
                i++;
        }
    }
};
